### PR TITLE
Fixed Matx::solve function for non-square matrixes

### LINKED ISSUE #11888: Error in Matx::solve for non-square matrices
##### System information (version)

- OpenCV => latest

##### Detailed description
See the following code

https://github.com/opencv/opencv/blob/3165baa1f152878b49c984384aa61c430a686768/modules/core/include/opencv2/core/matx.hpp#L183

`solve` is supported for non-square `matx`, i.e., `m!=n`.

Its implementation is

https://github.com/opencv/opencv/blob/3165baa1f152878b49c984384aa61c430a686768/modules/core/include/opencv2/core/operations.hpp#L208-L222

Pay attention to the line
https://github.com/opencv/opencv/blob/3165baa1f152878b49c984384aa61c430a686768/modules/core/include/opencv2/core/operations.hpp#L214

`Matx_FastSolveOp` is defined as

https://github.com/opencv/opencv/blob/3165baa1f152878b49c984384aa61c430a686768/modules/core/include/opencv2/core/operations.hpp#L121-L133

which requires a square matrix `const Matx<_Tp, m, m>& a`

##### Steps to reproduce

The following code will result in a compilation error

```.cpp
    cv::Matx<float, 3, 2> A(
            2, 1,
            3, 1,
            9, 1
    );
    cv::Vec<float, 3> b(4, 5, 7);

    auto x = A.solve(b, cv::DecompTypes::DECOMP_QR);
```

Error log is as follows:

```
........../include/opencv2/core/operations.hpp: In instantiation of ‘cv::Matx<_Tp, n, l> cv::Matx<_Tp, m, n>::solve(const cv::Matx<_Tp, m, l>&, int) const [with int l = 1; _Tp = float; int m = 3; int n = 2]’:
........../include/opencv2/core/matx.hpp:895:30:   required from ‘cv::Vec<_Tp, n> cv::Matx<_Tp, m, n>::solve(const cv::Vec<_Tp, m>&, int) const [with _Tp = float; int m = 3; int n = 2]’

........../include/opencv2/core/operations.hpp:214:57: error: no match for call to ‘(cv::internal::Matx_FastSolveOp<float, 3, 1>) (const cv::Matx<float, 3, 2>&, const cv::Matx<float, 3, 1>&, cv::Matx<float, 2, 1>&, int&)’
         ok = cv::internal::Matx_FastSolveOp<_Tp, m, l>()(*this, rhs, x, method);
                                                         ^
........../include/opencv2/core/operations.hpp:152:10: note: candidate: bool cv::internal::Matx_FastSolveOp<_Tp, 3, 1>::operator()(const cv::Matx<_Tp, 3, 3>&, const cv::Matx<T2, 3, 1>&, cv::Matx<T2, 3, 1>&, int) const [with _Tp = float]
     bool operator()(const Matx<_Tp, 3, 3>& a, const Matx<_Tp, 3, 1>& b,
          ^
........../include/opencv2/core/operations.hpp:152:10: note:   no known conversion for argument 1 from ‘const cv::Matx<float, 3, 2>’ to ‘const cv::Matx<float, 3, 3>&’
```

A workaround is to use
https://github.com/opencv/opencv/blob/3165baa1f152878b49c984384aa61c430a686768/modules/core/include/opencv2/core/operations.hpp#L215-L219

### PR DESCRIPTION
**Pull request by terfendail**
resolves #11888

### This pullrequest changes
Fixed compilation for `Matx::solve()` function with non-square matrixes

### CODE REVIEW

**Comment by alalek** [comment]
Please add test for this to avoid future regressions.

File: modules/core/include/opencv2/core/operations.hpp [hunk 1]
```diff
@@ -210,8 +210,8 @@ Matx<_Tp, n, l> Matx<_Tp, m, n>::solve(const Matx<_Tp, m, l>& rhs, int method) c
 {
     Matx<_Tp, n, l> x;
     bool ok;
-    if( method == DECOMP_LU || method == DECOMP_CHOLESKY )
-        ok = cv::internal::Matx_FastSolveOp<_Tp, m, l>()(*this, rhs, x, method);
+    if( m == n && (method == DECOMP_LU || method == DECOMP_CHOLESKY) )
```
Comment (csukuangfj) [inline modules/core/include/opencv2/core/operations.hpp]: I think this does not solve the problem and it will cause the same compilation error.

Comment (csukuangfj) [reply modules/core/include/opencv2/core/operations.hpp]: At compile time, the compiler won't check the conditions inside `if`.

**Comment by terfendail** [comment]
@csukuangfj @alalek Added test for the issue and reworked the fix

**Pull request merged**